### PR TITLE
promote CRD to v1

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind api-change
/kind feature

**What this PR does / why we need it**:
Promotes CRD types to v1. See the release note for changes

**Does this PR introduce a user-facing change?**:
```release-note
The `CustomResourceDefinition` API type is promoted to `apiextensions.k8s.io/v1` with the following changes:
* Use of the new `default` feature in validation schemas is limited to v1
* `spec.scope` is no longer defaulted to `Namespaced` and must be explicitly specified
* `spec.version` is removed; use `spec.versions` instead
* `spec.validation` is removed; use `spec.versions[*].schema` instead
* `spec.subresources` is removed; use `spec.versions[*].subresources` instead
* `spec.additionalPrinterColumns` is removed; use `spec.versions[*].additionalPrinterColumns` instead
* `spec.conversion.webhookClientConfig` is moved to `spec.conversion.webhook.clientConfig`
* `spec.conversion.conversionReviewVersions` is moved to `spec.conversion.webhook.conversionReviewVersions`
* `spec.versions[*].schema.openAPIV3Schema` is now required when creating v1 CustomResourceDefinitions
* `spec.preserveUnknownFields: true` is disallowed when creating v1 CustomResourceDefinitions; it must be specified within schema definitions as `x-kubernetes-preserve-unknown-fields: true`
* In `additionalPrinterColumns` items, the `JSONPath` field was renamed to `jsonPath` (fixes https://github.com/kubernetes/kubernetes/issues/66531)
The `apiextensions.k8s.io/v1beta1` version of `CustomResourceDefinition` is deprecated and will no longer be served in v1.19.
```

/sig api-machinery
/priority important-soon
/cc @jpbetz @sttts @roycaihw

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 1]
```diff
@@ -255,6 +255,17 @@ func validateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
 	allErrs = append(allErrs, validateCustomResourceDefinitionValidation(spec.Validation, mustBeStructural, hasAnyStatusEnabled(spec), opts, fldPath.Child("validation"))...)
 	allErrs = append(allErrs, ValidateCustomResourceDefinitionSubresources(spec.Subresources, fldPath.Child("subresources"))...)
 
+	// If no top-level schema is provided, ensure all individual versions define a schema
+	// v1beta1 defaults missing schemas to a minimal schema for compatibility.
+	// v1+ requires API clients to provide a schema.
+	if spec.Validation == nil {
+		for i, v := range spec.Versions {
+			if v.Schema == nil {
```
Comment (roycaihw) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: this is weird because people can create CRDs with empty schema
```
  versions:
  - name: v1
    schema: {}
    served: true
    storage: true
```

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go [hunk 2]
```diff
@@ -72,6 +73,38 @@ func SetDefaults_CustomResourceDefinitionSpec(obj *CustomResourceDefinitionSpec)
 	if obj.PreserveUnknownFields == nil {
 		obj.PreserveUnknownFields = utilpointer.BoolPtr(true)
 	}
+
+	// See if we need to default the schema.
+	// CRDs that specify spec.preserveUnknownFields=false are required to provide their own schemas.
+	if obj.Validation == nil && *obj.PreserveUnknownFields {
+		// Set up a minimal default
+		defaultSchema := &CustomResourceValidation{OpenAPIV3Schema: &JSONSchemaProps{Type: "object", XPreserveUnknownFields: utilpointer.BoolPtr(true)}}
```
Comment (roycaihw) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go]: This will break v1beta1 CRD with non-structural schema
https://github.com/kubernetes/kubernetes/blob/f4e39afea082afac7bf6f6936c34e012b7cb3c25/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L180-L182

perhaps we should scope `mustBeStructural` to each version

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types.go [hunk 3]
```diff
@@ -57,8 +57,9 @@ type CustomResourceDefinitionSpec struct {
 	// preserveUnknownFields disables pruning of object fields which are not
 	// specified in the OpenAPI schema. apiVersion, kind, metadata and known
 	// fields inside metadata are always preserved.
-	// Defaults to true in v1beta and will default to false in v1.
-	PreserveUnknownFields *bool `json:"preserveUnknownFields,omitempty" protobuf:"varint,10,opt,name=preserveUnknownFields"`
+	// This field is deprecated in favor of setting `x-preserve-unknown-fields` to true in `spec.versions[*].schema`.
```
Comment (roycaihw) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types.go]: nit: s/`spec.versions[*].schema`/`spec.versions[*].schema.openAPIV3Schema`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/install/install.go [hunk 4]
```diff
@@ -27,5 +28,6 @@ import (
 func Install(scheme *runtime.Scheme) {
 	utilruntime.Must(apiextensions.AddToScheme(scheme))
 	utilruntime.Must(v1beta1.AddToScheme(scheme))
-	utilruntime.Must(scheme.SetVersionPriority(v1beta1.SchemeGroupVersion))
+	utilruntime.Must(v1.AddToScheme(scheme))
+	utilruntime.Must(scheme.SetVersionPriority(v1beta1.SchemeGroupVersion, v1.SchemeGroupVersion))
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/install/install.go]: This was to prefer encoding to v1beta1 for storage

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion.go [hunk 5]
```diff
@@ -71,3 +71,104 @@ func Convert_v1_JSON_To_apiextensions_JSON(in *JSON, out *apiextensions.JSON, s
...
 }
+
+func Convert_apiextensions_CustomResourceDefinitionSpec_To_v1_CustomResourceDefinitionSpec(in *apiextensions.CustomResourceDefinitionSpec, out *CustomResourceDefinitionSpec, s conversion.Scope) error {
+	if err := autoConvert_apiextensions_CustomResourceDefinitionSpec_To_v1_CustomResourceDefinitionSpec(in, out, s); err != nil {
+		return err
+	}
+
+	if len(out.Versions) == 0 && len(in.Version) > 0 {
+		// no versions were specified, and a version name was specified
+		out.Versions = []CustomResourceDefinitionVersion{{Name: in.Version, Served: true, Storage: true}}
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion.go]: correct. This is to allow the internal client to convert to v1 correctly

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion_test.go [hunk 6]
```diff
@@ -18,13 +18,395 @@ package v1
...
+				},
+			},
+		},
+		{
+			Name: "v1 to internal, no versions",
+			In:   &CustomResourceDefinition{},
+			Out:  &apiextensions.CustomResourceDefinition{},
+			ExpectOut: &apiextensions.CustomResourceDefinition{
+				Spec: apiextensions.CustomResourceDefinitionSpec{
+					PreserveUnknownFields: pointer.BoolPtr(false),
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion_test.go]: We default to false and want to avoid serializing false values

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go]: good point. Will add a test and stew on what to do here. cc @sttts

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: that's neat. will stew on that.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/install/install.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (roycaihw) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/install/install.go]: I thought the storage version was controlled by `EncodeVersioner`, which we manually set v1beta1 for [CRD](https://github.com/kubernetes/kubernetes/blob/aecc1e829c54969c2c959bd9a7f374c3539d7980/cmd/kube-apiserver/app/apiextensions.go#L66). I see generally the storage version is inferred from the [Schema](https://github.com/kubernetes/kubernetes/blob/aecc1e829c54969c2c959bd9a7f374c3539d7980/staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go#L271) and then gets fed to `EncodeVersioner`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: yeah, will require both schema and openapiv3 schema to be set

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go]: we can require schema.openAPIV3Schema for create via v1, and require it for update via v1 if all versions in the existing object already have it

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go]: will drop the defaulting for v1beta1

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go]: dropped the defaulting for v1beta1, used ratcheting validation to require schemas in v1

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: fixed

File: cmd/kube-apiserver/app/apiextensions.go [hunk 7]
```diff
@@ -62,8 +63,8 @@ func createAPIExtensionsConfig(
 	// copy the etcd options so we don't mutate originals.
 	etcdOptions := *commandOptions.Etcd
 	etcdOptions.StorageConfig.Paging = utilfeature.DefaultFeatureGate.Enabled(features.APIListChunking)
-	etcdOptions.StorageConfig.Codec = apiextensionsapiserver.Codecs.LegacyCodec(v1beta1.SchemeGroupVersion)
-	etcdOptions.StorageConfig.EncodeVersioner = runtime.NewMultiGroupVersioner(v1beta1.SchemeGroupVersion, schema.GroupKind{Group: v1beta1.GroupName})
+	etcdOptions.StorageConfig.Codec = apiextensionsapiserver.Codecs.LegacyCodec(v1beta1.SchemeGroupVersion, v1.SchemeGroupVersion)
+	etcdOptions.StorageConfig.EncodeVersioner = runtime.NewMultiGroupVersioner(v1beta1.SchemeGroupVersion, schema.GroupKind{Group: v1beta1.GroupName}, schema.GroupKind{Group: v1.GroupName})
```
[the remark this answers is not in the document]
Comment (liggitt) [reply cmd/kube-apiserver/app/apiextensions.go]: fixed

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/install/install.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/install/install.go]: discovery/kubectl are controlled by the priority set in apiservices. you're right that the encode versioner controls storage (and the etcd integration test verifies we're storing in the expected version).

This controls what version someone using this scheme would get if they asked the scheme to encode an internal version and didn't specify a target external version. Sticking with v1beta1 for the first release v1 is available is reasonable.

[the remark this answers is not in the document]
Comment (roycaihw) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/install/install.go]: Thanks for the explanation

**Comment by liggitt** [comment]
@roycaihw thanks for the detailed review. comments addressed.

**Comment by roycaihw** [comment]
I feel we are missing the following [from the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190425-structural-openapi.md#api-compatibility-plan)
> 1. Updating a schema in a v1 CRD to a non-structural schema is only allowed if the original schema was non-structural already.
> 2. Non-structural CRDs only allow v1 schema updates if a structural schema is already set in the existing persisted object.

I found the second point hard to parse

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (roycaihw) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion_test.go]: I see. That's how we hide the field as we planned in the KEP

**Comment by liggitt** [comment]
> 1. Updating a schema in a v1 CRD to a non-structural schema is only allowed if the original schema was non-structural already.

added a commit and tests to do that (`CRD v1: require structural schema for v1`)

> 2. Non-structural CRDs only allow v1 schema updates if a structural schema is already set in the existing persisted object.

pretty sure that's a typo, will update

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 8]
```diff
@@ -158,10 +172,22 @@ func validateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
 
 	allErrs = append(allErrs, validateEnumStrings(fldPath.Child("scope"), string(spec.Scope), []string{string(apiextensions.ClusterScoped), string(apiextensions.NamespaceScoped)}, true)...)
 
-	mustBeStructural := false
+	// disabling pruning requires structural schemas
```
Comment (roycaihw) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: s/disabling/enabling

**Review by roycaihw — COMMENTED** [review]
/lgtm

/hold
for API ack from @smarterclayton

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go (hunk 8 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: fixed

**Comment by liggitt** [comment]
fixed comment typo, retagging

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types.go [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+	Reason string `json:"reason,omitempty" protobuf:"bytes,4,opt,name=reason"`
+	// Human-readable message indicating details about last transition.
+	// +optional
+	Message string `json:"message,omitempty" protobuf:"bytes,5,opt,name=message"`
+}
+
+// CustomResourceDefinitionStatus indicates the state of the CustomResourceDefinition
+type CustomResourceDefinitionStatus struct {
+	// Conditions indicate state for particular aspects of a CustomResourceDefinition
+	Conditions []CustomResourceDefinitionCondition `json:"conditions" protobuf:"bytes,1,opt,name=conditions"`
```
Comment (roycaihw) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types.go]: just an FYI I'm reviving https://github.com/kubernetes/kubernetes/pull/64996. The v1 API change could be done in a followup PR

**Comment by liggitt** [comment]
rebased on https://github.com/kubernetes/kubernetes/pull/79587

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types.go [hunk 10]
```diff
@@ -38,7 +38,7 @@ type CustomResourceDefinitionSpec struct {
 	Group string `json:"group" protobuf:"bytes,1,opt,name=group"`
 	// Names are the names used to describe this custom resource
 	Names CustomResourceDefinitionNames `json:"names" protobuf:"bytes,3,opt,name=names"`
-	// Scope indicates whether this resource is cluster or namespace scoped.  Default is namespaced
+	// Scope indicates whether this resource is cluster or namespace scoped.
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types.go]: I realized the godoc on this type doesn't follow our rules.  Can you file a follow up issue for 1.16 to get the godoc in the right naming form?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion.go [hunk 11]
```diff
@@ -71,3 +71,104 @@ func Convert_v1_JSON_To_apiextensions_JSON(in *JSON, out *apiextensions.JSON, s
...
+		}
+		if validationIdentical && !apiequality.Semantic.DeepEqual(v.Schema, validation) {
+			validationIdentical = false
+		}
+		if additionalPrinterColumnsIdentical && !apiequality.Semantic.DeepEqual(v.AdditionalPrinterColumns, additionalPrinterColumns) {
+			additionalPrinterColumnsIdentical = false
+		}
+	}
+
+	// If they are, set the top-level fields and clear the per-version fields
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion.go]: do we already have internal validation that ensures that this condition holds (one or the other)?

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion.go]: yes. https://github.com/kubernetes/kubernetes/blob/release-1.15/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L183-L192

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion_test.go [hunk 12]
```diff
@@ -18,13 +18,395 @@ package v1
...
+			},
+			Out: &CustomResourceDefinition{},
+			ExpectOut: &CustomResourceDefinition{
+				Spec: CustomResourceDefinitionSpec{
+					Versions: []CustomResourceDefinitionVersion{{Name: "v1", Served: true, Storage: true}},
+				},
+			},
+		},
+		{
+			Name: "internal to v1, multiple versions",
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion_test.go]: I assume there's an identical variant of this for v1beta1?  i.e. units guard this already in other versions?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go [hunk 13]
```diff
@@ -1159,6 +1161,25 @@ func validateAPIApproval(newCRD, oldCRD *apiextensions.CustomResourceDefinition,
 	}
 }
 
+func validatePreserveUnknownFields(crd, oldCRD *apiextensions.CustomResourceDefinition, requestGV schema.GroupVersion) field.ErrorList {
+	if requestGV == v1beta1.SchemeGroupVersion {
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: *twitch*

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion_test.go (hunk 12 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion_test.go]: internal and v1beta1 types match fields, so the conversion is just auto-generated.

v1beta1 defaulting ensures that empty version/versions are defaulted:
https://github.com/kubernetes/kubernetes/blob/32dc42ed34eb5ca97bd8fa4544d9a91d6b05a310/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go#L52-L63

validation ensures that spec.version matches spec.versions[0].name:
https://github.com/kubernetes/kubernetes/blob/32dc42ed34eb5ca97bd8fa4544d9a91d6b05a310/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L236-L238

and that spec.versions are specified:
https://github.com/kubernetes/kubernetes/blob/32dc42ed34eb5ca97bd8fa4544d9a91d6b05a310/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L185-L191

https://github.com/kubernetes/kubernetes/blob/32dc42ed34eb5ca97bd8fa4544d9a91d6b05a310/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L229-L231

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion_test.go]: and there's extensive validation and defaulting units for v1beta1

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go (hunk 13 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go]: yay compatibility

**Comment by liggitt** [comment]
as discussed, made the `require<ValidationAspect>()` helper functions take both the request version and any old persisted data, to make it clear that compatibility with existing data was being preserved. will squash fixup commits after review

File: staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/resources.go [hunk 14]
```diff
@@ -275,6 +276,63 @@ func CreateNewCustomResourceDefinition(crd *apiextensionsv1beta1.CustomResourceD
 	return crd, nil
 }
 
+// CreateNewV1CustomResourceDefinitionWatchUnsafe creates the CRD and makes sure
+// the apiextension apiserver has installed the CRD. But it's not safe to watch
+// the created CR. Please call CreateNewV1CustomResourceDefinition if you need to
+// watch the CR.
+func CreateNewV1CustomResourceDefinitionWatchUnsafe(v1CRD *apiextensionsv1.CustomResourceDefinition, apiExtensionsClient clientset.Interface) (*apiextensionsv1.CustomResourceDefinition, error) {
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/resources.go]: Strangely, the addition of Unsafe makes me feel less safe.

Comment (smarterclayton) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/resources.go]: Aren't these supposed to be in a separate bug fix?

Comment (smarterclayton) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/resources.go]: Are you pre-fixing this for the general issue?  Or it only applies here?

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/resources.go]: this is just a copy of the existing v1beta1 method to let you create a v1 version. it is fine if you have a test that doesn't use watch (this just skips waiting for watch caches to fill)

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/resources.go]: it's not really a bug, if you list/watch while the watch cache is establishing, you can get a 410 from the first watch request... a normal client will just re-list/re-watch, but deterministic watch tests hard fail.

**Comment by smarterclayton** [comment]
I think I follow everything non-mechanical, I'm sure I missed something but the tests make me feel comfortable that at least the individual pieces compose.

LGTM, squash your review commits and let's drop it.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types.go (hunk 10 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types.go]: https://github.com/kubernetes/kubernetes/issues/81519

**Comment by liggitt** [comment]
squashed/tagged per https://github.com/kubernetes/kubernetes/pull/79604#issuecomment-522083286
